# Movebank Data Analysis Attempt

After downloading the data, I found that the package came with two csv files. First, we can import the two tables - reference and Argos. To do so, we will have to import pandas (and numpy, for easier use later).

In [1]:
import pandas as pd
import numpy as np

Read the reference table

In [2]:
whale_ref = pd.read_csv("../data/Blue_whales_reference.csv")
whale_ref.head()

tag-id         animal-id           animal-taxon  \
0   1993CA-ST6-10823  1993CA-Bmu-10823  Balaenoptera musculus   
1   1993CA-ST6-10833  1993CA-Bmu-10833  Balaenoptera musculus   
2   1993CA-ST6-00834  1993CA-Bmu-00834  Balaenoptera musculus   
3   1993CA-ST6-10836  1993CA-Bmu-10836  Balaenoptera musculus   
4  1994CA-ST10-10821  1994CA-Bmu-10821  Balaenoptera musculus   

            deploy-on-date          deploy-off-date animal-life-stage  \
0  1993-08-28 18:20:00.000  1993-09-01 22:35:12.000             adult   
1  1993-08-28 18:54:00.000  1993-08-28 21:45:56.000             adult   
2  1993-08-29 01:13:00.000  1993-09-05 13:41:06.000             adult   
3  1993-08-31 23:40:00.000  1993-09-04 16:35:40.000             adult   
4  1994-09-13 20:31:00.000  1994-09-14 00:23:34.000             adult   

  animal-sex attachment-type  deploy-on-latitude  deploy-on-longitude  \
0                    implant              37.012             -122.412   
1                    implant              37.022             -122.415   
2                    implant              37.058             -122.433   
3                    implant              37.198             -122.773   
4                    implant              37.600             -123.000   

  deployment-id manipulation-type            study-site tag-manufacturer-name  \
0  1993CA-10823              none     Pt. Ano Nuevo, CA         Telonics, Inc   
1  1993CA-10833              none     Pt. Ano Nuevo, CA         Telonics, Inc   
2  1993CA-00834              none     Pt. Ano Nuevo, CA         Telonics, Inc   
3  1993CA-10836              none      Pigeon Point, CA         Telonics, Inc   
4  1994CA-10821              none  Farallon Islands, CA         Telonics, Inc   

  tag-model tag-readout-method  
0       ST6          satellite  
1       ST6          satellite  
2       ST6          satellite  
3       ST6          satellite  
4      ST10          satellite

In [3]:
whale_ref.shape

(143, 16)

The reference table records the information of each blue whale monitore. It is of size 143 rows x 16 columns. Let's figure out what each column is about.

In [4]:
whale_ref.columns

Index(['tag-id', 'animal-id', 'animal-taxon', 'deploy-on-date',
       'deploy-off-date', 'animal-life-stage', 'animal-sex', 'attachment-type',
       'deploy-on-latitude', 'deploy-on-longitude', 'deployment-id',
       'manipulation-type', 'study-site', 'tag-manufacturer-name', 'tag-model',
       'tag-readout-method'],
      dtype='object')

Combined with the column information provided in the README.txt file, we can find a few columns that provide informations that will be useful to us 
- *amimal-id*: An individual identifier for the subjects monitored
- *deploy-on-date* & *deploy-off-date*: the start and end time of monitoring
- *deploy-on-latitude* & *deploy-on-longitutde*: the start location of monitoring
- *study-site*: name of deployment site/facility

Before we downsize the table to a more concise version, let's check if the some standard measures (i.e. other less interesting columns) are uniform for each subject recorded.

In [5]:
whale_ref.groupby(['animal-life-stage','animal-taxon','attachment-type','manipulation-type','tag-readout-method']).count()

tag-id  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method           
adult             Balaenoptera musculus implant         none              satellite              143   

                                                                                              animal-id  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method              
adult             Balaenoptera musculus implant         none              satellite                 143   

                                                                                              deploy-on-date  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method                   
adult             Balaenoptera musculus implant         none              satellite                      143   

                                                                                              deploy-off-date  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method                    
adult             Balaenoptera musculus implant         none              satellite                       143   

                                                                                              animal-sex  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method               
adult             Balaenoptera musculus implant         none              satellite                  143   

                                                                                              deploy-on-latitude  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method                       
adult             Balaenoptera musculus implant         none              satellite                          143   

                                                                                              deploy-on-longitude  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method                        
adult             Balaenoptera musculus implant         none              satellite                           143   

                                                                                              deployment-id  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method                  
adult             Balaenoptera musculus implant         none              satellite                     143   

                                                                                              study-site  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method               
adult             Balaenoptera musculus implant         none              satellite                  143   

                                                                                              tag-manufacturer-name  \
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method                          
adult             Balaenoptera musculus implant         none              satellite                             143   

                                                                                              tag-model  
animal-life-stage animal-taxon          attachment-type manipulation-type tag-readout-method             
adult             Balaenoptera musculus implant         none              satellite                 143

In [6]:
whale_ref.groupby(['tag-manufacturer-name','tag-model']).count()

tag-id  animal-id  animal-taxon  \
tag-manufacturer-name   tag-model                                    
Telonics, Inc           ST10           15         15            15   
                        ST15          109        109           109   
                        ST21            6          6             6   
                        ST6            12         12            12   
Wildlife Computers, Inc MK10            1          1             1   

                                   deploy-on-date  deploy-off-date  \
tag-manufacturer-name   tag-model                                    
Telonics, Inc           ST10                   15               15   
                        ST15                  109              109   
                        ST21                    6                6   
                        ST6                    12               12   
Wildlife Computers, Inc MK10                    1                1   

                                   animal-life-stage  animal-sex  \
tag-manufacturer-name   tag-model                                  
Telonics, Inc           ST10                      15          15   
                        ST15                     109         109   
                        ST21                       6           6   
                        ST6                       12          12   
Wildlife Computers, Inc MK10                       1           1   

                                   attachment-type  deploy-on-latitude  \
tag-manufacturer-name   tag-model                                        
Telonics, Inc           ST10                    15                  15   
                        ST15                   109                 109   
                        ST21                     6                   6   
                        ST6                     12                  12   
Wildlife Computers, Inc MK10                     1                   1   

                                   deploy-on-longitude  deployment-id  \
tag-manufacturer-name   tag-model                                       
Telonics, Inc           ST10                        15             15   
                        ST15                       109            109   
                        ST21                         6              6   
                        ST6                         12             12   
Wildlife Computers, Inc MK10                         1              1   

                                   manipulation-type  study-site  \
tag-manufacturer-name   tag-model                                  
Telonics, Inc           ST10                      15          15   
                        ST15                     109         109   
                        ST21                       6           6   
                        ST6                       12          12   
Wildlife Computers, Inc MK10                       1           1   

                                   tag-readout-method  
tag-manufacturer-name   tag-model                      
Telonics, Inc           ST10                       15  
                        ST15                      109  
                        ST21                        6  
                        ST6                        12  
Wildlife Computers, Inc MK10                        1

Although most of the housekeeping columns have only one entry for all the rows (subjects), __the tag-model used for each subject is not exactly the same__ - most of which are ST15, and a few are of a slightly different model. __Will this create some problem for the data collected?__ We will find out later!

Now let's clean up our reference table for further use. 

In [7]:
whale_ref_clean = whale_ref[["animal-id","deploy-on-date","deploy-off-date","deploy-on-latitude",
                             "deploy-on-longitude","study-site","tag-manufacturer-name","tag-model"]]
whale_ref_clean.head(3)

animal-id           deploy-on-date          deploy-off-date  \
0  1993CA-Bmu-10823  1993-08-28 18:20:00.000  1993-09-01 22:35:12.000   
1  1993CA-Bmu-10833  1993-08-28 18:54:00.000  1993-08-28 21:45:56.000   
2  1993CA-Bmu-00834  1993-08-29 01:13:00.000  1993-09-05 13:41:06.000   

   deploy-on-latitude  deploy-on-longitude         study-site  \
0              37.012             -122.412  Pt. Ano Nuevo, CA   
1              37.022             -122.415  Pt. Ano Nuevo, CA   
2              37.058             -122.433  Pt. Ano Nuevo, CA   

  tag-manufacturer-name tag-model  
0         Telonics, Inc       ST6  
1         Telonics, Inc       ST6  
2         Telonics, Inc       ST6

In [8]:
whale_ref_clean.dtypes

animal-id                 object
deploy-on-date            object
deploy-off-date           object
deploy-on-latitude       float64
deploy-on-longitude      float64
study-site                object
tag-manufacturer-name     object
tag-model                 object
dtype: object

In [9]:
whale_ref_clean['deploy-on-date'] = pd.to_datetime(whale_ref_clean['deploy-on-date'])
whale_ref_clean['deploy-off-date'] = pd.to_datetime(whale_ref_clean['deploy-off-date'])
whale_ref_clean.dtypes

/Users/s.n.zhou/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/Users/s.n.zhou/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


animal-id                        object
deploy-on-date           datetime64[ns]
deploy-off-date          datetime64[ns]
deploy-on-latitude              float64
deploy-on-longitude             float64
study-site                       object
tag-manufacturer-name            object
tag-model                        object
dtype: object

In [10]:
whale_ref_clean.insert(3,'monitor-duration', whale_ref_clean['deploy-off-date'] - whale_ref_clean['deploy-on-date'])
whale_ref_clean.head(3)

animal-id      deploy-on-date     deploy-off-date monitor-duration  \
0  1993CA-Bmu-10823 1993-08-28 18:20:00 1993-09-01 22:35:12  4 days 04:15:12   
1  1993CA-Bmu-10833 1993-08-28 18:54:00 1993-08-28 21:45:56  0 days 02:51:56   
2  1993CA-Bmu-00834 1993-08-29 01:13:00 1993-09-05 13:41:06  7 days 12:28:06   

   deploy-on-latitude  deploy-on-longitude         study-site  \
0              37.012             -122.412  Pt. Ano Nuevo, CA   
1              37.022             -122.415  Pt. Ano Nuevo, CA   
2              37.058             -122.433  Pt. Ano Nuevo, CA   

  tag-manufacturer-name tag-model  
0         Telonics, Inc       ST6  
1         Telonics, Inc       ST6  
2         Telonics, Inc       ST6

In [21]:
whale_ref_clean.dtypes

animal-id                         object
deploy-on-date            datetime64[ns]
deploy-off-date           datetime64[ns]
monitor-duration         timedelta64[ns]
deploy-on-latitude               float64
deploy-on-longitude              float64
study-site                        object
tag-manufacturer-name             object
tag-model                         object
dtype: object

In [29]:
whale_ref_clean.sort_values('monitor-duration').head(5)

animal-id      deploy-on-date     deploy-off-date  \
104  2006CA-Bmu-00825 2006-09-12 22:33:00 2006-09-13 00:51:40   
1    1993CA-Bmu-10833 1993-08-28 18:54:00 1993-08-28 21:45:56   
4    1994CA-Bmu-10821 1994-09-13 20:31:00 1994-09-14 00:23:34   
8    1994CA-Bmu-10824 1994-09-14 22:00:00 1994-09-15 04:03:58   
127  2007CA-Bmu-00827 2007-09-09 20:51:00 2007-09-10 05:35:10   

    monitor-duration  deploy-on-latitude  deploy-on-longitude  \
104         02:18:40              34.056             -120.598   
1           02:51:56              37.022             -122.415   
4           03:52:34              37.600             -123.000   
8           06:03:58              37.745             -123.125   
127         08:44:10              34.122             -120.048   

                    study-site    tag-manufacturer-name tag-model  
104      San Miguel Island, CA            Telonics, Inc      ST15  
1            Pt. Ano Nuevo, CA            Telonics, Inc       ST6  
4         Farallon Islands, CA            Telonics, Inc      ST10  
8         Farallon Islands, CA            Telonics, Inc      ST10  
127  Santa Barbara Channel, CA  Wildlife Computers, Inc      MK10

Now let's move on to the juicer table! Repeat the same procedure done to the reference table.

In [11]:
whale_data = pd.read_csv("../data/Blue_whales_Argos.csv")
whale_data.head(3)

event-id  visible                timestamp  location-long  location-lat  \
0  9018011221     True  1993-08-29 01:13:00.000       -122.433        37.058   
1  9018011222     True  1993-08-28 18:20:00.000       -122.412        37.012   
2  9018011223     True  1993-08-28 18:54:00.000       -122.415        37.022   

   argos:best-level  argos:calcul-freq  argos:iq  argos:lat1  argos:lat2  \
0               NaN                NaN       NaN      37.058         NaN   
1               NaN                NaN       NaN      37.012         NaN   
2               NaN                NaN       NaN      37.022         NaN   

                         ...                         argos:lon1  argos:lon2  \
0                        ...                           -122.433         NaN   
1                        ...                           -122.412         NaN   
2                        ...                           -122.415         NaN   

   argos:nb-mes  argos:nb-mes-120  manually-marked-outlier sensor-type  \
0           NaN               NaN                      NaN         gps   
1           NaN               NaN                      NaN         gps   
2           NaN               NaN                      NaN         gps   

  individual-taxon-canonical-name tag-local-identifier  \
0           Balaenoptera musculus     1993CA-ST6-00834   
1           Balaenoptera musculus     1993CA-ST6-10823   
2           Balaenoptera musculus     1993CA-ST6-10833   

  individual-local-identifier  \
0            1993CA-Bmu-00834   
1            1993CA-Bmu-10823   
2            1993CA-Bmu-10833   

                                          study-name  
0  Blue whales Eastern North Pacific 1993-2008 - ...  
1  Blue whales Eastern North Pacific 1993-2008 - ...  
2  Blue whales Eastern North Pacific 1993-2008 - ...  

[3 rows x 21 columns]

In [12]:
whale_data.shape

(16249, 21)

In [13]:
whale_data.columns

Index(['event-id', 'visible', 'timestamp', 'location-long', 'location-lat',
       'argos:best-level', 'argos:calcul-freq', 'argos:iq', 'argos:lat1',
       'argos:lat2', 'argos:lc', 'argos:lon1', 'argos:lon2', 'argos:nb-mes',
       'argos:nb-mes-120', 'manually-marked-outlier', 'sensor-type',
       'individual-taxon-canonical-name', 'tag-local-identifier',
       'individual-local-identifier', 'study-name'],
      dtype='object')

Combined with the column information provided in the README.txt file, we can find a few columns that provide informations that will be useful to us 
- *timestamp*: The time point that a sensory measurement was taken
- *locaetion-long* & *location-lat*: the start location of monitoring
- *individual-local-identifier*: same as "animal-id" in reference table
- *manually-marked-outlier*: marked TRUE if "visible" marked FALSE

Some other technical measure associated with satellite signal receiving process that may be helpful for us to measure/visualize the credulity of data: 
- *argos:best-level*: Best signal strength
- *argos:calcul-freq*: Calculated frequency
- *argos:iq*: indicates transmitter oscillator frequency drift between two satellite passe
- *argos:nb-mes-120*: The number of messages received by the satellite at a signal strength greater than -120 decibels
- *sensor-type*: type of tracking sensor used (also appeared in reference table)



After identifying what we need and do not need, let us do some clean up and re-naming.

In [14]:
whale_data_clean = whale_data[['individual-local-identifier','timestamp','location-lat', 'location-long', 
                               'manually-marked-outlier','argos:best-level', 'argos:calcul-freq', 'argos:iq',
                               'argos:nb-mes','argos:nb-mes-120', 'sensor-type']]
whale_data_clean.columns = ['animal-id','timestamp','location-lat', 'location-long', 
                               'outlier','argos:best-level', 'argos:calcul-freq', 'argos:iq',
                               'argos:no-mes-rec','argos:no-mes-rec-120', 'sensor-type']
whale_data_clean.head()

animal-id                timestamp  location-lat  location-long  \
0  1993CA-Bmu-00834  1993-08-29 01:13:00.000        37.058       -122.433   
1  1993CA-Bmu-10823  1993-08-28 18:20:00.000        37.012       -122.412   
2  1993CA-Bmu-10833  1993-08-28 18:54:00.000        37.022       -122.415   
3  1993CA-Bmu-10836  1993-08-31 23:40:00.000        37.198       -122.773   
4  1994CA-Bmu-10820  1994-09-13 21:28:00.000        37.687       -123.102   

  outlier  argos:best-level  argos:calcul-freq  argos:iq  argos:no-mes-rec  \
0     NaN               NaN                NaN       NaN               NaN   
1     NaN               NaN                NaN       NaN               NaN   
2     NaN               NaN                NaN       NaN               NaN   
3     NaN               NaN                NaN       NaN               NaN   
4     NaN               NaN                NaN       NaN               NaN   

   argos:no-mes-rec-120 sensor-type  
0                   NaN         gps  
1                   NaN         gps  
2                   NaN         gps  
3                   NaN         gps  
4                   NaN         gps

In [15]:
whale_data.groupby(["manually-marked-outlier","individual-local-identifier"]).count()

event-id  visible  \
manually-marked-outlier individual-local-identifier                      
True                    1993CA-Bmu-00834                    1        1   
                        1993CA-Bmu-10823                    3        3   
                        1993CA-Bmu-10833                    1        1   
                        1994CA-Bmu-10822                    3        3   
                        1994CA-Bmu-10823                    1        1   
                        1994CA-Bmu-10828                    2        2   
                        1994CA-Bmu-10829                    1        1   
                        1995CA-Bmu-00841                    1        1   
                        1995CA-Bmu-04172                    5        5   
                        1995CA-Bmu-04173                    7        7   
                        1995CA-Bmu-23040                    9        9   
                        1995CA-Bmu-23041                    1        1   
                        1995CA-Bmu-23042                    6        6   
                        1995CA-Bmu-23043                    1        1   
                        1998CA-Bmu-00823                    8        8   
                        1998CA-Bmu-00824                    1        1   
                        1998CA-Bmu-00826                    6        6   
                        1998CA-Bmu-00827                    1        1   
                        1998CA-Bmu-00833                    5        5   
                        1998CA-Bmu-04174                    7        7   
                        1998CA-Bmu-04175                    7        7   
                        1998CA-Bmu-10823                    6        6   
                        1999CA-Bmu-00830                    2        2   
                        1999CA-Bmu-00832                    1        1   
                        1999CA-Bmu-00834                    3        3   
                        1999CA-Bmu-00837                    1        1   
                        1999CA-Bmu-00838                    5        5   
                        1999CA-Bmu-02082                    3        3   
                        1999CA-Bmu-04172                    7        7   
                        1999CA-Bmu-04176                    8        8   
...                                                       ...      ...   
                        2005CA-Bmu-10826                    1        1   
                        2005CA-Bmu-10828                    1        1   
                        2005CA-Bmu-10829                   10       10   
                        2005CA-Bmu-10830                    1        1   
                        2006CA-Bmu-00832                    5        5   
                        2006CA-Bmu-04175                    1        1   
                        2006CA-Bmu-10831                    3        3   
                        2006CA-Bmu-23036                    8        8   
                        2007CA-Bmu-00831                    4        4   
                        2007CA-Bmu-00835                    1        1   
                        2007CA-Bmu-00836                    3        3   
                        2007CA-Bmu-00847                    1        1   
                        2007CA-Bmu-01385                    1        1   
                        2007CA-Bmu-04172                    2        2   
                        2007CA-Bmu-04174                    3        3   
                        2007CA-Bmu-05801                    4        4   
                        2007CA-Bmu-10836                   10       10   
                        2007CA-Bmu-10838                    2        2   
                        2007CA-Bmu-23043                    4        4   
                        2008CA-Bmu-00825                    5        5   
                        2008CA-Bmu-00830                    9        9   
                        2008CA-Bmu-00840                 

In [16]:
whale_data.dtypes

event-id                             int64
visible                               bool
timestamp                           object
location-long                      float64
location-lat                       float64
argos:best-level                   float64
argos:calcul-freq                  float64
argos:iq                           float64
argos:lat1                         float64
argos:lat2                         float64
argos:lc                            object
argos:lon1                         float64
argos:lon2                         float64
argos:nb-mes                       float64
argos:nb-mes-120                   float64
manually-marked-outlier             object
sensor-type                         object
individual-taxon-canonical-name     object
tag-local-identifier                object
individual-local-identifier         object
study-name                          object
dtype: object

In [17]:
whale_data[[]]

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]

[16249 rows x 0 columns]